File: solutions/valid-perfect-square/valid-perfect-square.go
// @Title: Valid Perfect Square
// @Author: colinjxc
// @Date: 2022-03-05T22:20:25+08:00
// @URL: https://leetcode-cn.com/problems/valid-perfect-square


func isPerfectSquare(num int) bool {
    lo, hi := int64(0), int64(num)
    var square int64 = 0 
    for lo <= hi {
        mid := int64(lo + (hi - lo)/2)
        square = mid * mid
        if square > int64(num) {
            hi = mid -1
        } else if square < int64(num) {
            lo = mid + 1
        } else {
            return true
        }
    }
    return false
}

File: solutions/valid-perfect-square/valid-perfect-square.cpp
// @Title: Valid Perfect Square
// @Author: colinjxc
// @Date: 2022-03-05T22:14:22+08:00
// @URL: https://leetcode-cn.com/problems/valid-perfect-square


class Solution {
public:
    bool isPerfectSquare(int num) {
        long square = 1, n = 1;
        while (square <= num) {
            square = n*n;
            if (square == num) return true;
            n++;
        }
        return false;
    }
};
